### PR TITLE
remove unused variable from scope according final built instructions

### PR DESCRIPTION
**Pull request by CtfGo**
Some variables are eliminated by optimized passes(such as OpFusion), we can filter out them according to arguments of the built instructions, and erase them from the scope to avoid unnecessary buffer allocation

### CODE REVIEW

File: cinn/hlir/framework/graph_compiler.cc [hunk 1]
```diff
@@ -856,6 +859,47 @@ std::vector<std::unique_ptr<Instruction>> GraphCompiler::BuildInstructions() {
...
+                 var_names.end(),
+                 std::inserter(invalid_variables, invalid_variables.end()),
+                 [](const auto& name_view) { return std::string(name_view.data()); });
+
+  // erase used variable names
+  auto erase_used_name_fn = [&invalid_variables](const std::string& var_name) { invalid_variables.erase(var_name); };
+
+  auto exclude_arguments_fn = [&erase_used_name_fn](const std::vector<std::string>& args) {
+    std::for_each(args.begin(), args.end(), erase_used_name_fn);
+  };
```
Comment (wzzju) [inline cinn/hlir/framework/graph_compiler.cc]:
```suggestion
  auto exclude_arguments_fn = [&invalid_variables](const std::vector<std::string>& args) {
    std::for_each(args.begin(), args.end(), [&invalid_variables](const std::string& var_name) { invalid_variables.erase(var_name); });
  };
```

File: cinn/hlir/framework/graph_compiler.cc [hunk 2]
```diff
@@ -856,6 +859,47 @@ std::vector<std::unique_ptr<Instruction>> GraphCompiler::BuildInstructions() {
...
+    VLOG(3) << "Instruction-" << i << " eliminate " << unused_var_num - invalid_variables.size() << " used variables";
+    unused_var_num = invalid_variables.size();
+  }
+
+  VLOG(3) << "There are " << unused_var_num << " invalid variables to be removed from scope";
+  std::for_each(
+      invalid_variables.begin(), invalid_variables.end(), [& scope = this->scope_](const std::string& var_name) {
+        scope->EraseVar(var_name);
+        VLOG(3) << "Variable(" << var_name << ") is erased";
+      });
```
Comment (wzzju) [inline cinn/hlir/framework/graph_compiler.cc]:
```suggestion
  std::for_each(
      invalid_variables.begin(), invalid_variables.end(), [this](const std::string& var_name) {
        scope_->EraseVar(var_name);
        VLOG(3) << "Variable(" << var_name << ") is erased";
      });
```

**Comment by Superjomn** [comment]
I wonder why there are unused variables in scope, shouldn't all are used?

**Comment by CtfGo** [comment]
> I wonder why there are unused variables in scope, shouldn't all are used?

`BuildScope` create all variables of the initial graph not the optimized graph where some variables maybe eliminated by OpFusion pass currently

File: cinn/hlir/framework/graph_compiler.cc (hunk 1 above)
Comment (CtfGo) [reply cinn/hlir/framework/graph_compiler.cc]: take the advice

File: cinn/hlir/framework/graph_compiler.cc (hunk 2 above)
Comment (CtfGo) [reply cinn/hlir/framework/graph_compiler.cc]: take the advice

**Review by wzzju — APPROVED** [review]
LGTM

**Pull request merged**